In [1]:
#Import pandas, matplotlib.pyplot, and seaborn in the correct lines below
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns 
import os

Load and explore the Wildfire Data


In [2]:
file_path = "California_Fire_Incidents.xlsx"
wildfire_data = pd.read_excel(file_path, engine="openpyxl")

In [3]:
wildfire_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1636 entries, 0 to 1635
Data columns (total 40 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   AcresBurned           1633 non-null   float64
 1   Active                1636 non-null   bool   
 2   AdminUnit             1636 non-null   object 
 3   AirTankers            28 non-null     float64
 4   ArchiveYear           1636 non-null   int64  
 5   CalFireIncident       1636 non-null   bool   
 6   CanonicalUrl          1636 non-null   object 
 7   ConditionStatement    284 non-null    object 
 8   ControlStatement      105 non-null    object 
 9   Counties              1636 non-null   object 
 10  CountyIds             1636 non-null   object 
 11  CrewsInvolved         171 non-null    float64
 12  Dozers                123 non-null    float64
 13  Engines               191 non-null    float64
 14  Extinguished          1577 non-null   object 
 15  Fatalities           

# Check the data 

In [4]:
#Binary Classification is missing, and a lot of uncessary columns for the analysis
wildfire_data.head()

AcresBurned  Active                                          AdminUnit  \
0     257314.0   False  Stanislaus National Forest/Yosemite National Park   
1      30274.0   False  USFS Angeles National Forest/Los Angeles Count...   
2      27531.0   False  CAL FIRE Riverside Unit / San Bernardino Natio...   
3      27440.0   False                              Tahoe National Forest   
4      24251.0   False                       Ventura County Fire/CAL FIRE   

   AirTankers  ArchiveYear  CalFireIncident  \
0         NaN         2013             True   
1         NaN         2013             True   
2         NaN         2013             True   
3         NaN         2013            False   
4         NaN         2013             True   

                            CanonicalUrl  \
0         /incidents/2013/8/17/rim-fire/   
1  /incidents/2013/5/30/powerhouse-fire/   
2    /incidents/2013/7/15/mountain-fire/   
3    /incidents/2013/8/10/american-fire/   
4      /incidents/2013/5/2/springs-fire/   

                                  ConditionStatement ControlStatement  \
0                                                NaN              NaN   
1                                                NaN              NaN   
2                                                NaN              NaN   
3                                                NaN              NaN   
4  Acreage has been reduced based upon more accur...              NaN   

      Counties  ...                                     SearchKeywords  \
0     Tuolumne  ...  Rim Fire, Stanislaus National Forest, Yosemite...   
1  Los Angeles  ...  Powerhouse Fire, May 2013, June 2013, Angeles ...   
2    Riverside  ...  Mountain Fire, July 2013, Highway 243, Highway...   
3       Placer  ...  American Fire, August 2013, Deadwood Ridge, Fo...   
4      Ventura  ...  Springs Fire, May 2013, Highway 101, Camarillo...   

                Started     Status  StructuresDamaged StructuresDestroyed  \
0  2013-08-17T15:25:00Z  Finalized                NaN                 NaN   
1  2013-05-30T15:28:00Z  Finalized                NaN                 NaN   
2  2013-07-15T13:43:00Z  Finalized                NaN                 NaN   
3  2013-08-10T16:30:00Z  Finalized                NaN                 NaN   
4  2013-05-02T07:01:00Z  Finalized                6.0                10.0   

   StructuresEvacuated  StructuresThreatened  \
0                  NaN                   NaN   
1                  NaN                   NaN   
2                  NaN                   NaN   
3                  NaN                   NaN   
4                  NaN                   NaN   

                               UniqueId               Updated  WaterTenders  
0  5fb18d4d-213f-4d83-a179-daaf11939e78  2013-09-06T18:30:00Z           NaN  
1  bf37805e-1cc2-4208-9972-753e47874c87  2013-06-08T18:30:00Z           NaN  
2  a3149fec-4d48-427c-8b2c-59e8b79d59db  2013-07-30T18:00:00Z           NaN  
3  8213f5c7-34fa-403b-a4bc-da2ace6e6625  2013-08-30T08:00:00Z           NaN  
4  46731fb8-3350-4920-bdf7-910ac0eb715c  2013-05-11T06:30:00Z          11.0  

[5 rows x 40 columns]

In [5]:
wildfire_data.describe()  # Summarizes numerical columns

AcresBurned  AirTankers  ArchiveYear  CrewsInvolved      Dozers  \
count    1633.000000   28.000000  1636.000000     171.000000  123.000000   
mean     4589.443968    4.071429  2016.608802      11.561404    7.585366   
std     27266.337722    6.399818     1.845340      14.455633   14.028616   
min         0.000000    0.000000  2013.000000       0.000000    0.000000   
25%        35.000000    2.000000  2015.000000       2.500000    1.000000   
50%       100.000000    2.000000  2017.000000       6.000000    2.000000   
75%       422.000000    4.000000  2018.000000      13.500000    5.000000   
max    410203.000000   27.000000  2019.000000      82.000000   76.000000   

          Engines  Fatalities  Helicopters    Injuries     Latitude  \
count  191.000000   21.000000    84.000000  120.000000  1636.000000   
mean    23.565445    8.619048     5.357143    3.500000    37.203975   
std     41.004424   18.529642     7.265437    3.806231   135.401380   
min      0.000000    1.000000     0.000000    0.000000  -120.258000   
25%      5.000000    1.000000     1.000000    1.000000    34.165891   
50%     11.000000    3.000000     2.000000    3.000000    37.104065   
75%     24.000000    6.000000     5.000000    4.000000    39.086808   
max    256.000000   85.000000    29.000000   26.000000  5487.000000   

         Longitude  PercentContained  PersonnelInvolved  StructuresDamaged  \
count  1636.000000            1633.0         204.000000          67.000000   
mean   -108.082642             100.0         328.553922          67.970149   
std      37.006927               0.0         521.138789         155.771975   
min    -124.196290             100.0           0.000000           0.000000   
25%    -121.768358             100.0          55.000000           1.000000   
50%    -120.461560             100.0         151.500000           6.000000   
75%    -117.474073             100.0         350.000000          49.500000   
max     118.908200             100.0        3100.000000         783.000000   

       StructuresDestroyed  StructuresEvacuated  StructuresThreatened  \
count           175.000000                  0.0             30.000000   
mean            271.788571                  NaN            522.800000   
std            1557.255963                  NaN            739.586856   
min               0.000000                  NaN              0.000000   
25%               1.000000                  NaN              0.000000   
50%               7.000000                  NaN             45.000000   
75%              41.500000                  NaN           1043.750000   
max           18804.000000                  NaN           2600.000000   

       WaterTenders  
count    146.000000  
mean       7.815068  
std       12.719251  
min        1.000000  
25%        2.000000  
50%        4.000000  
75%        6.000000  
max       79.000000

In [6]:
wildfire_data.isnull().sum()

AcresBurned                3
Active                     0
AdminUnit                  0
AirTankers              1608
ArchiveYear                0
CalFireIncident            0
CanonicalUrl               0
ConditionStatement      1352
ControlStatement        1531
Counties                   0
CountyIds                  0
CrewsInvolved           1465
Dozers                  1513
Engines                 1445
Extinguished              59
Fatalities              1615
Featured                   0
Final                      0
FuelType                1624
Helicopters             1552
Injuries                1516
Latitude                   0
Location                   0
Longitude                  0
MajorIncident              0
Name                       0
PercentContained           3
PersonnelInvolved       1432
Public                     0
SearchDescription         17
SearchKeywords           203
Started                    0
Status                     0
StructuresDamaged       1569
StructuresDest

In [7]:
missing = pd.concat([wildfire_data.isnull().sum(), 100 * wildfire_data.isnull().mean()], axis=1)
missing.columns=['Count', '%']
missing.sort_values(by='Count', ascending=False) 

Count           %
StructuresEvacuated    1636  100.000000
FuelType               1624   99.266504
Fatalities             1615   98.716381
AirTankers             1608   98.288509
StructuresThreatened   1606   98.166259
StructuresDamaged      1569   95.904645
Helicopters            1552   94.865526
ControlStatement       1531   93.581907
Injuries               1516   92.665037
Dozers                 1513   92.481663
WaterTenders           1490   91.075795
CrewsInvolved          1465   89.547677
StructuresDestroyed    1461   89.303178
Engines                1445   88.325183
PersonnelInvolved      1432   87.530562
ConditionStatement     1352   82.640587
SearchKeywords          203   12.408313
Extinguished             59    3.606357
SearchDescription        17    1.039120
PercentContained          3    0.183374
AcresBurned               3    0.183374
Latitude                  0    0.000000
Started                   0    0.000000
Updated                   0    0.000000
UniqueId                  0    0.000000
AdminUnit                 0    0.000000
ArchiveYear               0    0.000000
Counties                  0    0.000000
CalFireIncident           0    0.000000
Status                    0    0.000000
Featured                  0    0.000000
Location                  0    0.000000
Final                     0    0.000000
Public                    0    0.000000
CanonicalUrl              0    0.000000
Active                    0    0.000000
CountyIds                 0    0.000000
MajorIncident             0    0.000000
Longitude                 0    0.000000
Name                      0    0.000000

For my project a lot of these columns that I am keeping does not have a lot of missing data which is good, the columns I am planning on keeping are `Latitude`,`Longtitude`,`ArchiveYear`,`AcresBurned` which is the only column with missing values, .18% which is relatively very low,`Location` although this column will be compared to `Counties` which ever one proves more insights is the one I'll, most likely it'll be `Counties` since this includes the exact county/region compared to `Location` which would require Natural Language Processing to draw insights. Lastly `WildfireOccured` is a column that I will be adding for the sake of this project. In terms of missing values my data set does not contain that many, other than `FuelType`,`StructuresEvacuated` which I'll get rid of since they are missing over 99% of their data.

In [8]:
wildfire_data['Location'].head()

0              3 miles east of Groveland along Hwy 120
1                              Angeles National Forest
2                Hwy 243 & Hwy 74 near Mountain Center
3              Deadwood Ridge, northeast of Foresthill
4    Southbound Highway 101 at Camarillo Springs Ro...
Name: Location, dtype: object

Check the unique values in the 'Counties' column so it can be mapped into broader categories

In [9]:
wildfire_data['Counties'].unique()

array(['Tuolumne', 'Los Angeles', 'Riverside', 'Placer', 'Ventura',
       'Fresno', 'Siskiyou', 'Humboldt', 'Tehama', 'Shasta', 'San Diego',
       'Kern', 'Sonoma', 'Contra Costa', 'Butte', 'Tulare',
       'Santa Barbara', 'Mariposa', 'Monterey', 'El Dorado',
       'San Bernardino', 'Plumas', 'Modoc', 'San Luis Obispo', 'Madera',
       'Inyo', 'Napa', 'San Benito', 'San Joaquin', 'Lake', 'Alameda',
       'Glenn', 'Yolo', 'Sacramento', 'Stanislaus', 'Solano', 'Merced',
       'Mendocino', 'Lassen', 'Amador', 'Yuba', 'Nevada', 'Santa Clara',
       'Calaveras', 'San Mateo', 'Orange', 'Colusa', 'Trinity',
       'Del Norte', 'Mono', 'Alpine', 'Sutter', 'Kings', 'Sierra',
       'Santa Cruz', 'Marin', 'Mexico', 'State of Oregon',
       'State of Nevada'], dtype=object)

In [10]:
wildfire_data['Counties'].nunique()

59

This will be used to perform EDA to see which county has the highest wildfire rates, and then it'll mapped into a broader region type into 4 different regions of 59 counties, to show which region is most susceptible to widlfires.

In [11]:
wildfire_data.describe()

AcresBurned  AirTankers  ArchiveYear  CrewsInvolved      Dozers  \
count    1633.000000   28.000000  1636.000000     171.000000  123.000000   
mean     4589.443968    4.071429  2016.608802      11.561404    7.585366   
std     27266.337722    6.399818     1.845340      14.455633   14.028616   
min         0.000000    0.000000  2013.000000       0.000000    0.000000   
25%        35.000000    2.000000  2015.000000       2.500000    1.000000   
50%       100.000000    2.000000  2017.000000       6.000000    2.000000   
75%       422.000000    4.000000  2018.000000      13.500000    5.000000   
max    410203.000000   27.000000  2019.000000      82.000000   76.000000   

          Engines  Fatalities  Helicopters    Injuries     Latitude  \
count  191.000000   21.000000    84.000000  120.000000  1636.000000   
mean    23.565445    8.619048     5.357143    3.500000    37.203975   
std     41.004424   18.529642     7.265437    3.806231   135.401380   
min      0.000000    1.000000     0.000000    0.000000  -120.258000   
25%      5.000000    1.000000     1.000000    1.000000    34.165891   
50%     11.000000    3.000000     2.000000    3.000000    37.104065   
75%     24.000000    6.000000     5.000000    4.000000    39.086808   
max    256.000000   85.000000    29.000000   26.000000  5487.000000   

         Longitude  PercentContained  PersonnelInvolved  StructuresDamaged  \
count  1636.000000            1633.0         204.000000          67.000000   
mean   -108.082642             100.0         328.553922          67.970149   
std      37.006927               0.0         521.138789         155.771975   
min    -124.196290             100.0           0.000000           0.000000   
25%    -121.768358             100.0          55.000000           1.000000   
50%    -120.461560             100.0         151.500000           6.000000   
75%    -117.474073             100.0         350.000000          49.500000   
max     118.908200             100.0        3100.000000         783.000000   

       StructuresDestroyed  StructuresEvacuated  StructuresThreatened  \
count           175.000000                  0.0             30.000000   
mean            271.788571                  NaN            522.800000   
std            1557.255963                  NaN            739.586856   
min               0.000000                  NaN              0.000000   
25%               1.000000                  NaN              0.000000   
50%               7.000000                  NaN             45.000000   
75%              41.500000                  NaN           1043.750000   
max           18804.000000                  NaN           2600.000000   

       WaterTenders  
count    146.000000  
mean       7.815068  
std       12.719251  
min        1.000000  
25%        2.000000  
50%        4.000000  
75%        6.000000  
max       79.000000

In [12]:
wildfire_data['Started'].dtype

dtype('O')

In [13]:
wildfire_data['Started'] = pd.to_datetime(wildfire_data['Started'], errors='coerce')

In [14]:
wildfire_data.dtypes

AcresBurned                         float64
Active                                 bool
AdminUnit                            object
AirTankers                          float64
ArchiveYear                           int64
CalFireIncident                        bool
CanonicalUrl                         object
ConditionStatement                   object
ControlStatement                     object
Counties                             object
CountyIds                            object
CrewsInvolved                       float64
Dozers                              float64
Engines                             float64
Extinguished                         object
Fatalities                          float64
Featured                               bool
Final                                  bool
FuelType                             object
Helicopters                         float64
Injuries                            float64
Latitude                            float64
Location                        

In [15]:
#As of now this is the only essential column to convert since I would need to compare this column with the weather data time
wildfire_data['Started'].dtype

datetime64[ns, UTC]

As of now the goal is to get rid of columns that are not necessary for this project so the data is clean, keeping only columns that are essential for making a model that wil be essential for binary classification of wildfires in different regions.

In [16]:
# Keep only the necessary columns
columns_to_keep = [
    'Latitude', 'Longitude', 'ArchiveYear', 'Started',
    'AcresBurned', 'Location', 'Counties'
]
# Filter the DataFrame
wildfire_data = wildfire_data[columns_to_keep]
# Display the result
wildfire_data.head()

Latitude   Longitude  ArchiveYear                   Started  AcresBurned  \
0  37.857000 -120.086000         2013 2013-08-17 15:25:00+00:00     257314.0   
1  34.585595 -118.423176         2013 2013-05-30 15:28:00+00:00      30274.0   
2  33.709500 -116.728850         2013 2013-07-15 13:43:00+00:00      27531.0   
3  39.120000 -120.650000         2013 2013-08-10 16:30:00+00:00      27440.0   
4   0.000000    0.000000         2013 2013-05-02 07:01:00+00:00      24251.0   

                                            Location     Counties  
0            3 miles east of Groveland along Hwy 120     Tuolumne  
1                            Angeles National Forest  Los Angeles  
2              Hwy 243 & Hwy 74 near Mountain Center    Riverside  
3            Deadwood Ridge, northeast of Foresthill       Placer  
4  Southbound Highway 101 at Camarillo Springs Ro...      Ventura

# Check if there are any duplicate rows in the dataset

In [17]:
wildfire_data[wildfire_data.duplicated()]

Empty DataFrame
Columns: [Latitude, Longitude, ArchiveYear, Started, AcresBurned, Location, Counties]
Index: []

### 🔍 Handling Duplicate Wildfire Records

After reviewing the dataset for duplicate entries, intially there were no exact duplicate rows. However, further inspection revealed multiple wildfire records with identical `Location` and `Started` timestamps, which is highly unlikely in real-world scenarios.

Since two wildfires cannot occur at the exact same location and start time, we treated these as redundant records. To ensure the integrity and accuracy of the data for modeling, we decided to remove duplicates based on the combination of `Location` and `Started`, keeping only the first occurrence of each.

This step helps eliminate noise and potential bias in our binary classification model, where we aim to predict wildfire occurrences based on temporal and spatial patterns.


In [18]:
wildfire_data[wildfire_data.duplicated(subset=['Latitude','Longitude','Started'], keep=False)].shape[0]

51

In [19]:
print("Before:", len(wildfire_data))

Before: 1636


In [20]:
wildfire_data = wildfire_data.drop_duplicates(
    subset=['Latitude', 'Longitude', 'Started', 'ArchiveYear', 'AcresBurned', 'Location'],
    keep='first'
)

In [21]:
print("After:", len(wildfire_data))

After: 1609


In [22]:
wildfire_data[wildfire_data.duplicated(subset=['Latitude','Longitude','Counties','AcresBurned'], keep=False)]

Latitude  Longitude  ArchiveYear                   Started  AcresBurned  \
416    0.00000    0.00000         2015 2015-07-02 17:10:00+00:00         25.0   
418    0.00000    0.00000         2015 2015-06-05 19:57:00+00:00         25.0   
594    0.00000    0.00000         2016 2016-08-06 15:00:00+00:00         25.0   
595    0.00000    0.00000         2016 2016-07-29 14:55:00+00:00         25.0   
611    0.00000    0.00000         2016 2016-07-22 14:11:00+00:00          0.0   
613    0.00000    0.00000         2016 2016-06-04 16:09:00+00:00          0.0   
1530  39.24678 -121.32399         2019 2019-06-25 13:51:09+00:00         80.0   
1531  39.24678 -121.32399         2019 2019-06-24 11:57:00+00:00         80.0   

                                               Location     Counties  
416    off Bucks Bar Cir & Buckaroo Trail near Somerset    El Dorado  
418         Hwy 20 near Lake County- Colusa County Line         Lake  
594           off of Hwy 50 at Riverton, west of Kyburz    El Dorado  
595          Off Ogulin Rd & Highway 53, Clearlake Oaks         Lake  
611         near 14 Fwy & Sand Canyon Rd, Santa Clarita  Los Angeles  
613   off of Old Topanga Canyon Road and Mulholland ...  Los Angeles  
1530  off Scott Forbes Rd, 5 miles east of Browns Va...         Yuba  
1531  off Scott Forbes Rd, 5 miles east of Browns Va...         Yuba

### Dropping Wildfire Records with Missing Geolocation

Rows where both latitude and longitude are 0.0 represent missing or invalid location data. Since geographic coordinates are essential for querying weather data and aligning wildfire occurrences to specific regions, these records are dropped to ensure model integrity.


In [23]:
wildfire_data = wildfire_data[~((wildfire_data['Latitude'] == 0.0) & (wildfire_data['Longitude'] == 0.0))]

In [24]:
len(wildfire_data)

1457

Row 1530 and 1531, looks a bit too identical, surely there can't a wildfire at the same place 2 different times burning the same amount of Acres, lets get rid of one of them.

In [25]:
wildfire_data[wildfire_data.duplicated(subset=['Latitude','Longitude','Counties'], keep=False)]

Latitude  Longitude  ArchiveYear                   Started  \
621   41.841000 -123.47400         2017 2017-08-15 07:55:00+00:00   
1050  41.841000 -123.47400         2017 2017-06-25 17:00:00+00:00   
1297  32.557566 -116.86630         2018 2018-07-17 14:05:00+00:00   
1530  39.246780 -121.32399         2019 2019-06-25 13:51:09+00:00   
1531  39.246780 -121.32399         2019 2019-06-24 11:57:00+00:00   
1597  32.557566 -116.86630         2019 2019-05-13 18:46:00+00:00   

      AcresBurned                                           Location  \
621       78698.0                        5 miles W of Happy Camp, CA   
1050          0.0  14 miles west of Etna, California in the Marbl...   
1297         28.0  north of the Meixcan border and east of the Ot...   
1530         80.0  off Scott Forbes Rd, 5 miles east of Browns Va...   
1531         80.0  off Scott Forbes Rd, 5 miles east of Browns Va...   
1597         22.0   off Tin Can Canyon along the border in Otay Mesa   

       Counties  
621    Siskiyou  
1050   Siskiyou  
1297  San Diego  
1530       Yuba  
1531       Yuba  
1597  San Diego

In [26]:
wildfire_data = wildfire_data.drop(index=1531)

In [27]:
wildfire_data[wildfire_data.duplicated(subset=['Latitude','Longitude','Counties'], keep=False)]

Latitude  Longitude  ArchiveYear                   Started  \
621   41.841000  -123.4740         2017 2017-08-15 07:55:00+00:00   
1050  41.841000  -123.4740         2017 2017-06-25 17:00:00+00:00   
1297  32.557566  -116.8663         2018 2018-07-17 14:05:00+00:00   
1597  32.557566  -116.8663         2019 2019-05-13 18:46:00+00:00   

      AcresBurned                                           Location  \
621       78698.0                        5 miles W of Happy Camp, CA   
1050          0.0  14 miles west of Etna, California in the Marbl...   
1297         28.0  north of the Meixcan border and east of the Ot...   
1597         22.0   off Tin Can Canyon along the border in Otay Mesa   

       Counties  
621    Siskiyou  
1050   Siskiyou  
1297  San Diego  
1597  San Diego

In [28]:
len(wildfire_data)

1456

Although some wildfires share identical latitude and longitude, further inspection shows they are distinct events based on different start dates, counties, or acres burned. Therefore, they are not considered true duplicates and will be retained in the dataset for modeling purposes.


In [29]:
wildfire_data[wildfire_data.duplicated(subset=['Latitude','Longitude'], keep=False)]

Latitude   Longitude  ArchiveYear                   Started  \
16    34.288877 -116.941311         2013 2013-05-01 12:38:00+00:00   
147   34.288877 -116.941311         2013 2013-04-20 17:30:00+00:00   
340   40.400196 -120.553075         2015 2015-09-09 14:30:00+00:00   
400   40.400196 -120.553075         2015 2015-09-09 14:12:00+00:00   
442   36.099740 -120.510570         2016 2016-08-09 13:08:00+00:00   
499   36.922440 -119.396190         2016 2016-07-11 14:30:00+00:00   
543   36.922440 -119.396190         2016 2016-07-13 15:52:00+00:00   
563   36.099740 -120.510570         2016 2016-08-09 17:30:00+00:00   
621   41.841000 -123.474000         2017 2017-08-15 07:55:00+00:00   
1050  41.841000 -123.474000         2017 2017-06-25 17:00:00+00:00   
1166  37.976380 -122.527500         2018 2018-06-09 14:45:00+00:00   
1297  32.557566 -116.866300         2018 2018-07-17 14:05:00+00:00   
1340  37.976380 -122.527500         2018 2018-06-09 15:30:00+00:00   
1422  39.289260 -121.859060         2019 2019-06-08 14:37:00+00:00   
1468  39.289260 -121.859060         2019 2019-06-17 17:50:00+00:00   
1597  32.557566 -116.866300         2019 2019-05-13 18:46:00+00:00   

      AcresBurned                                           Location  \
16         3166.0              Mias Canyon and Bluff Road in Banning   
147          30.0  west of Delamar Mountain, north of the communi...   
340         170.0              Squaw Valley Road and Sand Creek Road   
400          40.0                      Center Rd, east of Susanville   
442        7050.0    off Hwy 198 & Mineral Springs Rd, W of Coalinga   
499         188.0  off Watts Valley Rd. & Big Springs Ln, east of...   
543          78.0     near Lake Sherwood, southwest of Thousand Oaks   
563          45.0  off of Kugelman St and Hwy 53, community of Lo...   
621       78698.0                        5 miles W of Happy Camp, CA   
1050          0.0  14 miles west of Etna, California in the Marbl...   
1166        175.0               I-5 and Calgrove Blvd, Santa Clarita   
1297         28.0  north of the Meixcan border and east of the Ot...   
1340         12.0    Mission Avenue and C Street, City of San Rafael   
1422       1350.0  North Butte Rd & West Butte Rd, northeast of C...   
1468        207.0  off Highway 96 and Lumgrey Road, West of Gotts...   
1597         22.0   off Tin Can Canyon along the border in Otay Mesa   

            Counties  
16         Riverside  
147   San Bernardino  
340           Fresno  
400           Lassen  
442           Fresno  
499           Fresno  
543          Ventura  
563             Lake  
621         Siskiyou  
1050        Siskiyou  
1166     Los Angeles  
1297       San Diego  
1340           Marin  
1422          Sutter  
1468        Siskiyou  
1597       San Diego

**Now that we have cleaned up the dataset by removing redundant rows and columns that do not contribute meaningful insights, we are still missing a key component: the target variable for the binary classification model. Currently, all records represent actual wildfire events. To build a predictive model that determines whether a wildfire event will occur based on historical wildfire and weather data, we need a target variable.**

Create a new column, WildfireOccurred, where:

    `1` indicates that a wildfire event occurred.

    `0` will later represent non-wildfire events, either by simulation or by merging with an external dataset.

This target variable is crucial for setting up our binary classification problem.

In [30]:
wildfire_data['WildfireOccurred'] = 1

In [31]:
wildfire_data.head()

Latitude   Longitude  ArchiveYear                   Started  AcresBurned  \
0  37.857000 -120.086000         2013 2013-08-17 15:25:00+00:00     257314.0   
1  34.585595 -118.423176         2013 2013-05-30 15:28:00+00:00      30274.0   
2  33.709500 -116.728850         2013 2013-07-15 13:43:00+00:00      27531.0   
3  39.120000 -120.650000         2013 2013-08-10 16:30:00+00:00      27440.0   
5  37.279000 -119.318000         2013 2013-07-22 22:15:00+00:00      22992.0   

                                  Location     Counties  WildfireOccurred  
0  3 miles east of Groveland along Hwy 120     Tuolumne                 1  
1                  Angeles National Forest  Los Angeles                 1  
2    Hwy 243 & Hwy 74 near Mountain Center    Riverside                 1  
3  Deadwood Ridge, northeast of Foresthill       Placer                 1  
5           Seven miles north of Big Creek       Fresno                 1

Since we already have `Latitude` & `Longitude`, along with `Counties` and a broader region category `Region`, we have a lot of Geographical information, we can get rid of `Location` column, it is redundant.

In [32]:
wildfire_data.drop(columns=['Location'], inplace=True)

In [33]:
wildfire_data.head()

Latitude   Longitude  ArchiveYear                   Started  AcresBurned  \
0  37.857000 -120.086000         2013 2013-08-17 15:25:00+00:00     257314.0   
1  34.585595 -118.423176         2013 2013-05-30 15:28:00+00:00      30274.0   
2  33.709500 -116.728850         2013 2013-07-15 13:43:00+00:00      27531.0   
3  39.120000 -120.650000         2013 2013-08-10 16:30:00+00:00      27440.0   
5  37.279000 -119.318000         2013 2013-07-22 22:15:00+00:00      22992.0   

      Counties  WildfireOccurred  
0     Tuolumne                 1  
1  Los Angeles                 1  
2    Riverside                 1  
3       Placer                 1  
5       Fresno                 1

In [34]:
wildfire_data['Latitude'].min()

-120.258

In [35]:
wildfire_data['Latitude'].max()

5487.0

This is likely a data entry, its an invalid coordinate that is invalid coordinate

In [36]:
wildfire_data[wildfire_data['Latitude'] == 5487.0]

Latitude   Longitude  ArchiveYear                   Started  AcresBurned  \
528    5487.0 -120.841536         2016 2016-07-09 12:43:00+00:00         95.0   

       Counties  WildfireOccurred  
528  San Benito                 1

In [37]:
wildfire_data['Longitude'].min()

-124.19629

In [38]:
wildfire_data['Longitude'].max()

118.9082

In [39]:
wildfire_data[wildfire_data['Longitude'] == 118.9082]

Latitude  Longitude  ArchiveYear                   Started  AcresBurned  \
474    34.913   118.9082         2016 2016-07-08 11:15:00+00:00        554.0   

    Counties  WildfireOccurred  
474     Kern                 1

Turn all `Longitude` to fit in the bounded range of California, along with `Latitude`

In [40]:
# Flip positive longitudes to negative
wildfire_data['Longitude'] = wildfire_data['Longitude'].apply(lambda x: -abs(x))

In [41]:
wildfire_data[wildfire_data['Longitude'] == 118.9082]

Empty DataFrame
Columns: [Latitude, Longitude, ArchiveYear, Started, AcresBurned, Counties, WildfireOccurred]
Index: []

In [42]:
wildfire_data[wildfire_data['Longitude'] > 0 ]

Empty DataFrame
Columns: [Latitude, Longitude, ArchiveYear, Started, AcresBurned, Counties, WildfireOccurred]
Index: []

In [43]:
wildfire_data['ArchiveYear'].max()

2019

In [44]:
wildfire_data = wildfire_data[
    (wildfire_data['Latitude'] >= 32) & (wildfire_data['Latitude'] <= 42.5) &
    (wildfire_data['Longitude'] >= -124.5) & (wildfire_data['Longitude'] <= -114.0)
]

In [45]:
len(wildfire_data)

1443

In [46]:
wildfire_data['Latitude'].max()

42.048

In [47]:
wildfire_data[wildfire_data['Counties'].isin(['Mexico', 'State of Oregon', 'State of Nevada'])]

Latitude   Longitude  ArchiveYear                   Started  \
1421  32.573437 -116.658735         2019 2019-08-16 06:14:55+00:00   
1423  42.033000 -122.175000         2019 2019-08-09 23:08:25+00:00   
1424  39.631945 -119.769881         2019 2019-07-13 18:42:22+00:00   
1590  32.557546 -116.873955         2019 2019-08-25 16:15:15+00:00   

      AcresBurned         Counties  WildfireOccurred  
1421       1507.0           Mexico                 1  
1423       1301.0  State of Oregon                 1  
1424       1165.0  State of Nevada                 1  
1590         25.0           Mexico                 1

In [48]:
# Missing values in 'Started' and 'AcresBurned' columns
wildfire_data = wildfire_data.dropna(subset=['Started', 'AcresBurned'])

In [49]:
wildfire_data['ArchiveYear'].unique()

array([2013, 2014, 2015, 2016, 2017, 2018, 2019], dtype=int64)

In [50]:
wildfire_data = wildfire_data[wildfire_data['Started'].dt.year != 1969].reset_index(drop=True)

In [51]:
len(wildfire_data)

1434

In [52]:
#wildfire_data.to_csv('cleaned_wildfire_data.csv', index = False)

`AcrhiveYear` Will be extrated later after the simulation is done from the Started column

In [53]:
wildfire_data = wildfire_data.drop(columns=['ArchiveYear'])

In [54]:
wildfire_data = wildfire_data.reset_index(drop=True)

In [55]:
wildfire_data.head()

Latitude   Longitude                   Started  AcresBurned     Counties  \
0  37.857000 -120.086000 2013-08-17 15:25:00+00:00     257314.0     Tuolumne   
1  34.585595 -118.423176 2013-05-30 15:28:00+00:00      30274.0  Los Angeles   
2  33.709500 -116.728850 2013-07-15 13:43:00+00:00      27531.0    Riverside   
3  39.120000 -120.650000 2013-08-10 16:30:00+00:00      27440.0       Placer   
4  37.279000 -119.318000 2013-07-22 22:15:00+00:00      22992.0       Fresno   

   WildfireOccurred  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

Since the current dataset contains only records of actual wildfire evetns `WildfireOccured` = 1, we need to simulate realistic non-wildfire scenarioos `WildfireOccured` = 0 for binary classification. In this process I'll be generating locations and dates that did not expierence wildfires the timeframe and area covered in the orignal dataset.

In [85]:
# %pip install geopandas

In [57]:
import geopandas as gpd
gdf = gpd.read_file("California_Counties.geojson")

In [58]:
gdf.head()

OBJECTID              NAME  STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  \
0         1    Alameda County  California         06       001  06001   
1         2     Alpine County  California         06       003  06003   
2         3     Amador County  California         06       005  06005   
3         4      Butte County  California         06       007  06007   
4         5  Calaveras County  California         06       009  06009   

    Shape__Area  Shape__Length  \
0  3.084661e+09  448425.775672   
1  3.156006e+09  275565.131463   
2  2.562531e+09  359593.235353   
3  7.340033e+09  525974.550843   
4  4.356209e+09  367005.674344   

                                            geometry  
0  MULTIPOLYGON (((-122.19112 37.69298, -122.1910...  
1  POLYGON ((-119.90061 38.93077, -119.88726 38.9...  
2  POLYGON ((-120.07765 38.70889, -120.07771 38.7...  
3  POLYGON ((-121.40463 40.14665, -121.40489 40.1...  
4  POLYGON ((-120.07247 38.50916, -120.07254 38.4...

In [59]:
gdf['Counties'] = gdf['NAME'].str.replace(' County', '', regex=False)

In [60]:
gdf[['Counties', 'geometry']].head()

Counties                                           geometry
0    Alameda  MULTIPOLYGON (((-122.19112 37.69298, -122.1910...
1     Alpine  POLYGON ((-119.90061 38.93077, -119.88726 38.9...
2     Amador  POLYGON ((-120.07765 38.70889, -120.07771 38.7...
3      Butte  POLYGON ((-121.40463 40.14665, -121.40489 40.1...
4  Calaveras  POLYGON ((-120.07247 38.50916, -120.07254 38.4...

In [61]:
import random
from shapely.geometry import Point
from datetime import datetime, timedelta
import pandas as pd

# 1. Prep geometry + wildfire dates
valid_gdf      = gdf[['Counties','geometry']].copy()
wildfire_dates = set(wildfire_data['Started'].dt.date)

# 2. Helper to sample a point inside any (multi)polygon
def random_point_in_geom(geom):
    minx, miny, maxx, maxy = geom.bounds
    while True:
        x, y = random.uniform(minx, maxx), random.uniform(miny, maxy)
        p = Point(x, y)
        if geom.contains(p):
            return p

# 3. Define date window & total days
start_dt   = datetime(2013, 1, 1)
end_dt     = datetime(2019,12,31)
total_days = (end_dt - start_dt).days

# 4. Figure out which counties to skip entirely
skip_counties = {'Mexico', 'State of Oregon', 'State of Nevada'}

# 5. Count real wildfires per county (but drop the three we don’t want)
wf_counts = (
    wildfire_data.loc[~wildfire_data['Counties'].isin(skip_counties), 'Counties'].value_counts()
)

# 6. Simulate non-fire events proportional to real counts
sim_rows = []
for county, n_sim in wf_counts.items():
    # pull this county’s polygon
    geom = valid_gdf.loc[valid_gdf['Counties'] == county, 'geometry']
    if geom.empty:
        # should not happen now, since skip_counties removed the out‐of‐bounds ones
        continue
    geom = geom.iloc[0]

    for _ in range(n_sim):
        # a) random point inside that county
        pt = random_point_in_geom(geom)

        # b) random datetime not matching any real-fire date
        while True:
            cand = start_dt + timedelta(
                days   = random.randrange(total_days),
                hours  = random.randrange(24),
                minutes= random.randrange(60)
            )
            if cand.date() not in wildfire_dates:
                rand_dt = cand
                break

        # c) build the record
        sim_rows.append({
            "Latitude":         round(pt.y, 6),
            "Longitude":        round(pt.x, 6),
            "Started":          rand_dt.strftime("%Y-%m-%d %H:%M:%S+00:00"),
            "AcresBurned":      0,
            "Counties":         county,
            "WildfireOccurred": 0
        })

simulated_df = pd.DataFrame(sim_rows)

In [62]:
simulated_df.head()

Latitude   Longitude                    Started  AcresBurned   Counties  \
0  34.029136 -116.338586  2014-05-09 10:42:00+00:00            0  Riverside   
1  33.744879 -115.678002  2015-01-23 11:23:00+00:00            0  Riverside   
2  33.542878 -116.001602  2017-03-01 10:14:00+00:00            0  Riverside   
3  33.639541 -116.867235  2017-12-29 18:07:00+00:00            0  Riverside   
4  34.055974 -115.136930  2018-03-04 16:51:00+00:00            0  Riverside   

   WildfireOccurred  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [104]:
wildfire_data[wildfire_data['Counties'] == 'Los Angeles'].head(10)

Latitude   Longitude                   Started  AcresBurned  \
1    34.585595 -118.423176 2013-05-30 15:28:00+00:00      30274.0   
27   34.491490 -118.610044 2013-05-17 13:23:00+00:00        712.0   
46   34.149000 -117.908000 2013-09-23 18:00:00+00:00        268.0   
65   34.419284 -118.586880 2013-05-28 11:30:00+00:00        149.0   
155  34.170300 -117.881600 2014-01-16 05:50:00+00:00       1952.0   
174  34.267000 -117.867000 2014-08-17 14:30:00+00:00        274.0   
180  34.235000 -117.771000 2014-08-14 12:48:00+00:00        190.0   
230  34.361007 -118.535793 2015-06-24 13:22:00+00:00        415.0   
237  34.367000 -117.683000 2015-07-17 21:30:00+00:00        200.0   
316  34.310000 -118.335000 2016-11-14 05:45:00+00:00       4824.0   

        Counties  WildfireOccurred  
1    Los Angeles                 1  
27   Los Angeles                 1  
46   Los Angeles                 1  
65   Los Angeles                 1  
155  Los Angeles                 1  
174  Los Angeles                 1  
180  Los Angeles                 1  
230  Los Angeles                 1  
237  Los Angeles                 1  
316  Los Angeles                 1

In [103]:
simulated_df[simulated_df['Counties'] == 'Los Angeles'].head(10)

Latitude   Longitude                    Started  AcresBurned  \
726  34.255396 -117.670934  2013-09-12 06:25:00+00:00            0   
727  34.138108 -117.956211  2013-04-29 03:33:00+00:00            0   
728  34.447193 -117.996611  2014-05-31 10:10:00+00:00            0   
729  34.240509 -117.677685  2017-05-03 19:25:00+00:00            0   
730  34.274009 -117.959417  2013-03-30 00:45:00+00:00            0   
731  34.514834 -118.077695  2013-03-28 00:14:00+00:00            0   
732  34.797953 -118.013986  2015-05-08 19:57:00+00:00            0   
733  34.297403 -117.663813  2018-12-19 10:39:00+00:00            0   
734  34.542488 -117.988333  2016-04-09 17:28:00+00:00            0   
735  33.953754 -118.407491  2016-06-02 22:17:00+00:00            0   

        Counties  WildfireOccurred  
726  Los Angeles                 0  
727  Los Angeles                 0  
728  Los Angeles                 0  
729  Los Angeles                 0  
730  Los Angeles                 0  
731  Los Angeles                 0  
732  Los Angeles                 0  
733  Los Angeles                 0  
734  Los Angeles                 0  
735  Los Angeles                 0

In [65]:
wildfire_data[wildfire_data['Counties'] == 'Amador'].shape[0]

11

In [66]:
simulated_df[simulated_df['Counties'] == 'Amador'].shape[0]

11

In [67]:
merged_data = pd.concat([wildfire_data, simulated_df], ignore_index=True)

In [68]:
merged_data.head()

Latitude   Longitude                    Started  AcresBurned     Counties  \
0  37.857000 -120.086000  2013-08-17 15:25:00+00:00     257314.0     Tuolumne   
1  34.585595 -118.423176  2013-05-30 15:28:00+00:00      30274.0  Los Angeles   
2  33.709500 -116.728850  2013-07-15 13:43:00+00:00      27531.0    Riverside   
3  39.120000 -120.650000  2013-08-10 16:30:00+00:00      27440.0       Placer   
4  37.279000 -119.318000  2013-07-22 22:15:00+00:00      22992.0       Fresno   

   WildfireOccurred  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [106]:
len(merged_data)

2864

### Mapping Counties to Broader Regions

To simplify the analysis and make the wildfire prediction model more interpretable, we will be 
mapping all the counties into broader regions based on their geographical location within California.

The regions will be as follows:
- **Northern California**
- **Central California**
- **Southern California**
- **Western California**

We created a dictionary to map each county into one of these four regions. A new column, called **'Region'**, will help capture regional trends in wildfire occurrences, rather than treating every county separately, although we can defintely run analysis on that as well.


In [87]:
county_to_region = {
    # Northern California
    'Del Norte': 'Northern California', 'Humboldt': 'Northern California', 'Mendocino': 'Northern California', 
    'Lake': 'Northern California', 'Trinity': 'Northern California', 'Siskiyou': 'Northern California', 
    'Shasta': 'Northern California', 'Tehama': 'Northern California', 'Lassen': 'Northern California', 
    'Plumas': 'Northern California', 'Modoc': 'Northern California', 'Nevada': 'Northern California', 
    'Butte': 'Northern California', 'Glenn': 'Northern California',

    # Central California
    'Yolo': 'Central California', 'Sutter': 'Central California', 'Colusa': 'Central California', 
    'Yuba': 'Central California', 'Placer': 'Central California', 'El Dorado': 'Central California', 
    'Sierra': 'Central California', 'Alpine': 'Central California', 'Amador': 'Central California', 
    'Calaveras': 'Central California', 'Tuolumne': 'Central California', 'Mariposa': 'Central California', 
    'Mono': 'Central California', 'Stanislaus': 'Central California', 'San Joaquin': 'Central California', 
    'Merced': 'Central California', 'Madera': 'Central California', 'Fresno': 'Central California', 
    'Kings': 'Central California', 'Tulare': 'Central California', 'Sacramento': 'Central California',

    # Southern California
    'Kern': 'Southern California', 'San Luis Obispo': 'Southern California', 'Santa Barbara': 'Southern California', 
    'Ventura': 'Southern California', 'Los Angeles': 'Southern California', 'Orange': 'Southern California', 
    'Riverside': 'Southern California', 'San Bernardino': 'Southern California', 'San Diego': 'Southern California', 
    'Inyo': 'Southern California',

    # Western California
    'Sonoma': 'Western California', 'Napa': 'Western California', 'Marin': 'Western California', 
    'Contra Costa': 'Western California', 'Alameda': 'Western California', 'San Mateo': 'Western California', 
    'Santa Clara': 'Western California', 'Santa Cruz': 'Western California', 'Monterey': 'Western California', 
    'Solano': 'Western California', 'San Benito': 'Western California',

    # Special cases
    'Mexico': 'Southern California',
    'State of Oregon': 'Northern California',
    'State of Nevada': 'Northern California'
}

In [88]:
# Map counties to regions
merged_data['Region'] = merged_data['Counties'].map(county_to_region)

In [89]:
merged_data[merged_data['Region'].isnull()]['Counties'].unique()

array([], dtype=object)

In [ ]:
print(merged_data['Region'].value_counts())

Region
Southern California    960
Northern California    826
Central California     658
Western California     420
Name: count, dtype: int64


In [74]:
merged_data['Started'] = pd.to_datetime(merged_data['Started'])

In [75]:
merged_data['Date'] = merged_data['Started'].dt.date

In [76]:
merged_data[['Started', 'Date']].head()

Started        Date
0 2013-08-17 15:25:00+00:00  2013-08-17
1 2013-05-30 15:28:00+00:00  2013-05-30
2 2013-07-15 13:43:00+00:00  2013-07-15
3 2013-08-10 16:30:00+00:00  2013-08-10
4 2013-07-22 22:15:00+00:00  2013-07-22

In [77]:
#Extract the year from the 'Started' column and create a new column 'ArchiveYear'
merged_data['ArchiveYear'] = merged_data['Started'].dt.year

cols = list(merged_data.columns)
cols.insert(2, cols.pop(cols.index('ArchiveYear')))
merged_data = merged_data[cols]

Sort the actual wildfire occurences and non wildfire occurences in a chronologicall format

In [78]:
merged_data = merged_data.sort_values('Started').reset_index(drop=True)

In [79]:
merged_data.head()

Latitude   Longitude  ArchiveYear                   Started  AcresBurned  \
0  40.785933 -121.017469         2013 2013-01-01 08:14:00+00:00          0.0   
1  33.469231 -115.114638         2013 2013-01-02 16:51:00+00:00          0.0   
2  40.816333 -121.838142         2013 2013-01-04 15:07:00+00:00          0.0   
3  34.186179 -118.661168         2013 2013-01-06 17:30:00+00:00          0.0   
4  37.507826 -122.105244         2013 2013-01-08 05:47:00+00:00          0.0   

      Counties  WildfireOccurred               Region        Date  
0       Lassen                 0  Northern California  2013-01-01  
1    Riverside                 0  Southern California  2013-01-02  
2       Shasta                 0  Northern California  2013-01-04  
3  Los Angeles                 0  Southern California  2013-01-06  
4      Alameda                 0   Western California  2013-01-08

Now that our dataset is completely cleaned while retaining enough of the original dataset, combined with a simulated dataset for the sake of this binary classification project, in the next EDA I'll need to merge Open Meteo Weather API, for data analysis. Since weather patterns are correlated to wildfire occurrences, I'll be able to know exactly what weather patterns cause the highest spread of wildfire and wildfire occurrences based on regions. Our data wrangling is complete, in the next step, I'll be carefully analyzing these statistical patterns more closely.

In [105]:
merged_data.to_csv('cleaned_wildfire_data.csv', index = False)